### PR TITLE
Sq 8 2+azure devops

### PR DESCRIPTION
**Pull request by jcuzzi**
Manual merge of @Iloer's Azure DevOps PR into @mc1arke's sq-8_2-support branch (with some additions to get it all working).

- Add support for Azure DevOps PR Decoration

Notes: 

1. The following should be set in the scanner properties:
sonar.pullrequest.vsts.instanceUrl=https://[organization].visualstudio.com/
sonar.pullrequest.vsts.project=[Project]
sonar.pullrequest.vsts.repository=[Repository]

2. Ensure that the following option is **unchecked** in the Azure DevOps Branch Policy configuration, otherwise an exception will be thrown:
**"Reset status whenever there are new changes"**

### CODE REVIEW

**Comment by twaddell** [comment]
I can confirm this build works well in sonarqube version 8.3.1.34397
I've tested with a couple of PRs and comments were added in Azure Devops. I will leave this build running on our server and report back if any issues arise.

**Comment by btshft** [comment]
I get the following error on version 8.3.1.34397. Any ideas? 

<details>
<summary>Log</summary>

```
##[section]Starting: Scan - Analyze
==============================================================================
Task         : Run Code Analysis
Description  : Run scanner and upload the results to the SonarQube server.
Version      : 4.11.0
Author       : sonarsource
Help         : Version: 4.11.0. This task is not needed for Maven and Gradle projects since the scanner should be run as part of the build.

==============================================================================

[command]/usr/bin/dotnet /azp/agent/_work/_tasks/SonarQubePrepare_15b84ca1-b62f-4a2a-a403-89b77a063157/4.11.0/dotnet-sonar-scanner-msbuild/SonarScanner.MSBuild.dll end
SonarScanner for MSBuild 4.10
Using the .NET Core version of the Scanner for MSBuild
Post-processing started.
Calling the SonarQube Scanner...
INFO: Scanner configuration file: /azp/agent/_work/_tasks/SonarQubePrepare_15b84ca1-b62f-4a2a-a403-89b77a063157/4.11.0/dotnet-sonar-scanner-msbuild/sonar-scanner-4.4.0.2170/conf/sonar-scanner.properties
INFO: Project root configuration file: /azp/agent/_work/1/.sonarqube/out/sonar-project.properties
INFO: SonarScanner 4.4.0.2170
INFO: Java 11.0.7 Ubuntu (64-bit)
INFO: Linux 4.15.0-106-generic amd64
INFO: User cache: /root/.sonar/cache
INFO: Scanner configuration file: /azp/agent/_work/_tasks/SonarQubePrepare_15b84ca1-b62f-4a2a-a403-89b77a063157/4.11.0/dotnet-sonar-scanner-msbuild/sonar-scanner-4.4.0.2170/conf/sonar-scanner.properties
INFO: Project root configuration file: /azp/agent/_work/1/.sonarqube/out/sonar-project.properties
INFO: Analyzing on SonarQube server 8.3.1
INFO: Default locale: "en_US", source code encoding: "US-ASCII" (analysis is platform dependent)
INFO: Load global settings
INFO: Load global settings (done) | time=76ms
INFO: Server id: F86E3401-AXLisUehgVv8qPQEMlBu
INFO: User cache: /root/.sonar/cache
INFO: Load/download plugins
INFO: Load plugins index
INFO: Load plugins index (done) | time=57ms
INFO: Load/download plugins (done) | time=166ms
INFO: Process project properties
INFO: Process project properties (done) | time=60ms
INFO: Execute project builders
INFO: Execute project builders (done) | time=81ms
INFO: Project key: *********
INFO: Base dir: /azp/agent/_work/1/s/src
INFO: Working dir: /azp/agent/_work/1/.sonarqube/out/.sonar
INFO: Load project settings for component key: '*********'
INFO: Load project settings for component key: '*********' (done) | time=23ms
INFO: Load project branches
INFO: Load project branches (done) | time=17ms
INFO: Load project pull requests
INFO: Load project pull requests (done) | time=10ms
INFO: Load branch configuration
INFO: ------------------------------------------------------------------------
INFO: EXECUTION FAILURE
INFO: ------------------------------------------------------------------------
INFO: Total time: 2.151s
INFO: Final Memory: 15M/181M
INFO: ------------------------------------------------------------------------
##[error]ERROR: Error during SonarScanner execution
ERROR: Error during SonarScanner execution
##[error]java.lang.IllegalStateException: Unable to load component class org.sonar.scanner.scan.filesystem.InputComponentStore
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:66)
java.lang.IllegalStateException: Unable to load component class org.sonar.scanner.scan.filesystem.InputComponentStore
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:66)
##[error]at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:632)
	at org.picocontainer.parameters.BasicComponentParameter$1.resolveInstance(BasicComponentParameter.java:118)
	at org.picocontainer.parameters.ComponentParameter$1.resolveInstance(ComponentParameter.java:136)
	at org.picocontainer.injectors.SingleMemberInjector.getParameter(SingleMemberInjector.java:78)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:632)
	at org.picocontainer.parameters.BasicComponentParameter$1.resolveInstance(BasicComponentParameter.java:118)
	at org.picocontainer.parameters.ComponentParameter$1.resolveInstance(ComponentParameter.java:136)
	at org.picocontainer.injectors.SingleMemberInjector.getParameter(SingleMemberInjector.java:78)
##[error]at org.picocontainer.injectors.ConstructorInjector$CtorAndAdapters.getParameterArguments(ConstructorInjector.java:309)
	at org.picocontainer.injectors.ConstructorInjector$1.run(ConstructorInjector.java:335)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.ConstructorInjector.getComponentInstance(ConstructorInjector.java:364)
	at org.picocontainer.injectors.ConstructorInjector$CtorAndAdapters.getParameterArguments(ConstructorInjector.java:309)
	at org.picocontainer.injectors.ConstructorInjector$1.run(ConstructorInjector.java:335)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.ConstructorInjector.getComponentInstance(ConstructorInjector.java:364)
##[error]at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.getComponentInstance(AbstractInjectionFactory.java:56)
	at org.picocontainer.behaviors.AbstractBehavior.getComponentInstance(AbstractBehavior.java:64)
	at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.getComponentInstance(AbstractInjectionFactory.java:56)
	at org.picocontainer.behaviors.AbstractBehavior.getComponentInstance(AbstractBehavior.java:64)
##[error]at org.picocontainer.behaviors.Stored.getComponentInstance(Stored.java:91)
	at org.picocontainer.DefaultPicoContainer.instantiateComponentAsIsStartable(DefaultPicoContainer.java:1034)
	at org.picocontainer.behaviors.Stored.getComponentInstance(Stored.java:91)
	at org.picocontainer.DefaultPicoContainer.instantiateComponentAsIsStartable(DefaultPicoContainer.java:1034)
##[error]at org.picocontainer.DefaultPicoContainer.addAdapterIfStartable(DefaultPicoContainer.java:1026)
	at org.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:1003)
	at org.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:767)
	at org.picocontainer.DefaultPicoContainer.addAdapterIfStartable(DefaultPicoContainer.java:1026)
	at org.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:1003)
	at org.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:767)
##[error]at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:136)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
##[error]at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:141)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:141)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
##[error]at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:72)
	at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:72)
##[error]at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:66)
	at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:66)
##[error]at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy0.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
	at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.IllegalStateException: Unable to load component interface org.sonar.scanner.scan.branch.BranchConfiguration
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:66)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:632)
	at org.picocontainer.parameters.BasicComponentParameter$1.resolveInstance(BasicComponentParameter.java:118)
	at org.picocontainer.parameters.ComponentParameter$1.resolveInstance(ComponentParameter.java:136)
	at org.picocontainer.injectors.SingleMemberInjector.getParameter(SingleMemberInjector.java:78)
	at org.picocontainer.injectors.ConstructorInjector$CtorAndAdapters.getParameterArguments(ConstructorInjector.java:309)
	at org.picocontainer.injectors.ConstructorInjector$1.run(ConstructorInjector.java:335)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.ConstructorInjector.getComponentInstance(ConstructorInjector.java:364)
	at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.getComponentInstance(AbstractInjectionFactory.java:56)
	at org.picocontainer.behaviors.AbstractBehavior.getComponentInstance(AbstractBehavior.java:64)
	at org.picocontainer.behaviors.Stored.getComponentInstance(Stored.java:91)
	at org.picocontainer.DefaultPicoContainer.getInstance(DefaultPicoContainer.java:699)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:647)
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:64)
	... 34 more
Caused by: java.lang.NoSuchFieldError: SHORT
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.computeBranchType(CommunityBranchConfigurationLoader.java:147)
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.createBranchConfiguration(CommunityBranchConfigurationLoader.java:127)
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.load(CommunityBranchConfigurationLoader.java:96)
	at org.sonar.scanner.scan.branch.BranchConfigurationProvider.provide(BranchConfigurationProvider.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.picocontainer.injectors.MethodInjector.invokeMethod(MethodInjector.java:129)
	at org.picocontainer.injectors.MethodInjector.access$000(MethodInjector.java:39)
	at org.picocontainer.injectors.MethodInjector$2.run(MethodInjector.java:113)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.MethodInjector.decorateComponentInstance(MethodInjector.java:120)
	at org.picocontainer.injectors.CompositeInjector.decorateComponentInstance(CompositeInjector.java:58)
	at org.picocontainer.injectors.Reinjector.reinject(Reinjector.java:142)
	at org.picocontainer.injectors.ProviderAdapter.getComponentInstance(ProviderAdapter.java:96)
	at org.picocontainer.DefaultPicoContainer.getInstance(DefaultPicoContainer.java:699)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:647)
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:64)
	... 48 more
ERROR:
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy0.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
	at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.IllegalStateException: Unable to load component interface org.sonar.scanner.scan.branch.BranchConfiguration
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:66)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:632)
	at org.picocontainer.parameters.BasicComponentParameter$1.resolveInstance(BasicComponentParameter.java:118)
	at org.picocontainer.parameters.ComponentParameter$1.resolveInstance(ComponentParameter.java:136)
	at org.picocontainer.injectors.SingleMemberInjector.getParameter(SingleMemberInjector.java:78)
	at org.picocontainer.injectors.ConstructorInjector$CtorAndAdapters.getParameterArguments(ConstructorInjector.java:309)
	at org.picocontainer.injectors.ConstructorInjector$1.run(ConstructorInjector.java:335)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.ConstructorInjector.getComponentInstance(ConstructorInjector.java:364)
	at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.getComponentInstance(AbstractInjectionFactory.java:56)
	at org.picocontainer.behaviors.AbstractBehavior.getComponentInstance(AbstractBehavior.java:64)
	at org.picocontainer.behaviors.Stored.getComponentInstance(Stored.java:91)
	at org.picocontainer.DefaultPicoContainer.getInstance(DefaultPicoContainer.java:699)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:647)
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:64)
	... 34 more
Caused by: java.lang.NoSuchFieldError: SHORT
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.computeBranchType(CommunityBranchConfigurationLoader.java:147)
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.createBranchConfiguration(CommunityBranchConfigurationLoader.java:127)
	at com.github.mc1arke.sonarqube.plugin.scanner.CommunityBranchConfigurationLoader.load(CommunityBranchConfigurationLoader.java:96)
	at org.sonar.scanner.scan.branch.BranchConfigurationProvider.provide(BranchConfigurationProvider.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.picocontainer.injectors.MethodInjector.invokeMethod(MethodInjector.java:129)
	at org.picocontainer.injectors.MethodInjector.access$000(MethodInjector.java:39)
	at org.picocontainer.injectors.MethodInjector$2.run(MethodInjector.java:113)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.MethodInjector.decorateComponentInstance(MethodInjector.java:120)
	at org.picocontainer.injectors.CompositeInjector.decorateComponentInstance(CompositeInjector.java:58)
	at org.picocontainer.injectors.Reinjector.reinject(Reinjector.java:142)
	at org.picocontainer.injectors.ProviderAdapter.getComponentInstance(ProviderAdapter.java:96)
	at org.picocontainer.DefaultPicoContainer.getInstance(DefaultPicoContainer.java:699)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:647)
	at org.sonar.core.platform.ComponentContainer$ExtendedDefaultPicoContainer.getComponent(ComponentContainer.java:64)
	... 48 more
ERROR: 
##[error]The SonarQube Scanner did not complete successfully
The SonarQube Scanner did not complete successfully
##[error]18:42:17.722  Post-processing failed. Exit code: 1
18:42:17.722  Post-processing failed. Exit code: 1
##[error]The process '/usr/bin/dotnet' failed with exit code 1
```

</details>

**Comment by btshft** [comment]
Now I can confirm, this build works fine with SonarQube 8.3.1.34397 and Azure on-premise with REST API 5.1-preview.
The only thing I had to rebuild the plugin with the Azure REST API version 5.1-preview. So it would be nice to move version to plugin settings instead of hardcode constant.

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.Map;
+
+
+
+public class AzureDevOpsServerPullRequestDecorator implements PullRequestBuildStatusDecorator {
+
+    private String authorizationHeader;
+    public static final String AZURE_API_VERSION = "?api-version=6.0-preview.1";
```
Comment (alekssako) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java:52]: Adding this as a scanner property would be nice, not all versions of api included in all types of azure devops/tfs onpremises.

Comment (jcuzzi) [reply src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java:52]: I'm fine with that, but if there is a change in a version of the API that's not compatible with the code, then it won't work properly.  As is, it should work fine, but its not future proof...

Comment (jcuzzi) [reply src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java]: @alekssako @btshft
0a34144 adds the option to use the following scanner property to set the desired api version. Example:

sonar.pullrequest.vsts.apiVersion=5.1-preview.1

**Comment by alekssako** [comment]
Throwing error on `ce.log` :
```
Caused by: java.net.URISyntaxException: Illegal character in path at index 53: http://xxxxxxxxxxxxxx/tfs/xxxxxxxxx/DEVOPS IMPLEMENTATION/_apis/git/repositories/xxxxxxxxxxx/pullRequests/4004/statuses?api-version=5.1-preview.1
        at java.base/java.net.URI$Parser.fail(URI.java:2913)
        at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
        at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3166)
        at java.base/java.net.URI$Parser.parse(URI.java:3114)
        at java.base/java.net.URI.<init>(URI.java:600)
        at java.base/java.net.URI.create(URI.java:881)
        ... 24 common frames omitted
```

I think should fix spaces on project names or repository names. If having spaces it throws errors.

**Comment by jcuzzi** [comment]
@alekssako f203253 allows for spaces in the URI - it should encode the full URI before sending the request.

Thanks for the reviews.

**Comment by alekssako** [comment]
> @alekssako [f203253](https://github.com/mc1arke/sonarqube-community-branch-plugin/commit/f203253bc289a217acce702083095ce04d79a494) allows for spaces in the URI - it should encode the full URI before sending the request.
> 
> Thanks for the reviews.

Keep it up with the good work you`ve done.

I found out that in sonarqube "See the PR" link is referencing to the API instead of the PR itself on the azure devops.

Annotation 2020-07-09 102250


Can't find it in the code, but its something like :
`http://xxxxxxx/tfs/xxxxxx/ProjectName/_apis/git/repositories/RepoName/pullRequests/4004` this is referencing to the api,

Instead it should be something like : 
`http://xxxxxxx/tfs/xxxxxx/ProjectName/_git/RepoName/pullrequest/4004`

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+                LOGGER.error("sendPost: " + EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8));
+                throw new IllegalStateException("An error was returned in the response from the Azure DevOps server. See the previous log messages for details");
+            } else if (null != httpResponse) {
+                LOGGER.debug("sendPost: " + httpResponse.toString());
+                LOGGER.info("sendPost: " + successMessage);
+            }
+        }
+    }
+
+    private <T> T sendGet(String apiUrl, Class<T> type) throws IOException, MalformedURLException, URISyntaxException {
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java:360]: The `MalformedURLException` is a subtype of `IOException` so is covered by that `throws` so can be removed.

File: src/main/java/com/github/mc1arke/sonarqube/plugin/scanner/ScannerPullRequestPropertySensor.java [hunk 3]
```diff
@@ -42,6 +46,15 @@ public void describe(SensorDescriptor sensorDescriptor) {
...
     public void execute(SensorContext sensorContext) {
+        projectConfiguration.get(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_INSTANCE_URL).ifPresent(v -> sensorContext
+                .addContextProperty(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_INSTANCE_URL, v));
+        projectConfiguration.get(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PROJECT_ID).ifPresent(v -> sensorContext
+                .addContextProperty(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PROJECT_ID, v));
+        projectConfiguration.get(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PROJECT_URL).ifPresent(v -> sensorContext
+                .addContextProperty(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PROJECT_URL, v));
+        projectConfiguration.get(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PIPELINE_ID).ifPresent(v -> sensorContext
+                .addContextProperty(GitlabServerPullRequestDecorator.PULLREQUEST_GITLAB_PIPELINE_ID, v));
+
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/scanner/ScannerPullRequestPropertySensor.java:57]: Why have these been added? They seem unrelated to the intent of your change

File: src/main/java/com/github/mc1arke/sonarqube/plugin/scanner/ScannerPullRequestPropertySensor.java [hunk 4]
```diff
@@ -61,6 +74,35 @@ public void execute(SensorContext sensorContext) {
...
+                .addContextProperty(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_REPOSITORY_NAME, v));
+        projectConfiguration.get(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_BASE_BRANCH).ifPresent(v -> sensorContext
+                .addContextProperty(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_BASE_BRANCH, v));
+        projectConfiguration.get(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_BRANCH).ifPresent(v -> sensorContext
+                .addContextProperty(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_BRANCH, v));
+        projectConfiguration.get(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_PULLREQUEST_ID).ifPresent(v -> sensorContext
+                .addContextProperty(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_PULLREQUEST_ID, v));
+        
+        
+        Optional.ofNullable(system2.property(AzureDevOpsServerPullRequestDecorator.PULLREQUEST_AZUREDEVOPS_INSTANCE_URL)).ifPresent(
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/scanner/ScannerPullRequestPropertySensor.java:96]: Should these variables not be the ones that are [injected by Azure DevOps](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml), rather than custom-named ones?

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java [hunk 5]
```diff
@@ -121,6 +122,10 @@ public String getIssueUrl(String issueKey) {
         return qualityGate.getStatus();
     }
 
+    public Map<String,String> getScannerProperties() {
+        return scannerContext.getProperties();
+    }
+
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:128]: This only seems to have been introduced to allow logging of all the properties, which doesn't seem like a good use case. The logging should either happen in the PostAnalysisTask that triggers decoration, or be removed.

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+import java.util.Optional;
+import java.util.Map;
+
+
+
+public class AzureDevOpsServerPullRequestDecorator implements PullRequestBuildStatusDecorator {
+
+    private String authorizationHeader;    
+    private static final Logger LOGGER = Loggers.get(AzureDevOpsServerPullRequestDecorator.class);
+    public static final String API_VERSION_PREFIX = "?api-version=";
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java:58]: Please ensure you're following the Java Language Specification field ordering:
public static final
private static final
private final
private

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+
+    private String apiVersion = "6.0-preview.1";
+    private String azureUrl = "";
+    private String baseBranch = "";
+    private String branch = "";
+    private String pullRequestId = "";
+    private String azureRepositoryName = "";
+    private String azureProjectId = "";
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java:66]: These should all be local variables that get passed to any required methods, otherwise it will cause a race condition if multiple decorators are executed at the same time

**Comment by mc1arke** [comment]
Sorry, I've just merged the 8.2 support branch which has auto-closed your MR. Could you open a new one pointing at master please?

**Comment by jcuzzi** [comment]
Yeah, once I have time to make the requested changes, I'll create a new PR to master.

**Comment by jcuzzi** [comment]
> 
> 
> > @alekssako [f203253](https://github.com/mc1arke/sonarqube-community-branch-plugin/commit/f203253bc289a217acce702083095ce04d79a494) allows for spaces in the URI - it should encode the full URI before sending the request.
> > Thanks for the reviews.
> 
> @jcuzzi Keep it up with the good work you`ve done.
> 
> I found out that in sonarqube "See the PR" link is referencing to the API instead of the PR itself on the azure devops.
> 
> Annotation 2020-07-09 102250
> 
> Can't find it in the code, but its something like :
> `http://xxxxxxx/tfs/xxxxxx/ProjectName/_apis/git/repositories/RepoName/pullRequests/4004` this is referencing to the api,
> 
> Instead it should be something like :
> `http://xxxxxxx/tfs/xxxxxx/ProjectName/_git/RepoName/pullrequest/4004`

This should be fixed in #218
